### PR TITLE
vg/vgsvg: add switch to embed fonts to SVG plot

### LINKED ISSUE #703: vg/vgsvg: consider adding the ability to embed fonts in SVG
as evocated in #702, we could perhaps add an option to `vg/vgsvg.Canvas` (like we did for `vg/vgpdf.Canvas`) to embed fonts inside the SVG document.

the default should probably be to *not* embed fonts.

ex:
```svg
<svg xmlns="http://www.w3.org/2000/svg" width="450" height="150" font-size="24" text-anchor="middle">
    <defs>
        <style>
            @font-face{
                font-family:"Roboto Condensed";
                src:url(data:font/ttf;charset=utf-8;base64,your_base64_encoded_long_string) format("ttf"); 
                font-weight:normal;font-style:normal;
            }
            @font-face{
                font-family:"Open Sans";
                src:url(data:font/ttf;charset=utf-8;base64,your_base64_encoded_long_string) format("ttf"); 
                font-weight:normal;font-style:normal;
            }
            @font-face{
                font-family:"Anonymous Pro";
                src:url(data:font/ttf;charset=utf-8;base64,your_base64_encoded_long_string) format("ttf"); 
                font-weight:normal;font-style:normal;
            }
        </style>
    </defs>
        <text font-family="Roboto Condensed" x="190" y="32.92">
        This is a Roboto Condensed font
    </text>
    <text font-family="Open Sans" x="190" y="82.92">
        This is a Open Sans font
    </text>
    <text font-family="Anonymous Pro" x="190" y="132.92">
        This is a Anonymous Pro font
    </text>
</svg>
```

### PR DESCRIPTION
**Pull request by sbinet**
Fixes #703.

### CODE REVIEW

File: vg/vgsvg/vgsvg.go [hunk 1]
```diff
@@ -17,8 +17,10 @@ import (
 	"image/png"
 	"io"
 	"math"
+	"strings"
 
 	svgo "github.com/ajstarks/svgo"
+	stdfnt "golang.org/x/image/font"
 
 	"gonum.org/v1/plot/font"
 	"gonum.org/v1/plot/vg"
```
Comment (kortschak) [inline vg/vgsvg/vgsvg.go:26]: I think somewhere in this file we need documentation about how to ensure that the standard fonts are available on the system if they are not embedded.

Comment (sbinet) [reply vg/vgsvg/vgsvg.go:26]: you mean adding documentation (or a mechanism) to be able to use the `Times/Helvetica/Courier` fonts when embedding isn't enabled?

Comment (kortschak) [reply vg/vgsvg/vgsvg.go:26]: I think just documentation.

Comment (sbinet) [reply vg/vgsvg/vgsvg.go:26]: I've added an example instead.

**Comment by sbinet** [comment]
PTAL.

**Review by kortschak — COMMENTED** [review]
We still need docs for the case when the fonts aren't embedded, otherwise we are bound to see issues from people complaining about the effect that I show in [this comment](https://github.com/gonum/plot/pull/704#issuecomment-861015532).

File: vg/vgsvg/vgsvg.go [hunk 2]
```diff
@@ -408,6 +408,19 @@ func svgFamilyName(fnt font.Font) string {
 	if fnt.Variant != "" {
 		v += ", " + svgVariantName(fnt.Variant)
 	}
+
+	// add failover to "standard" fonts (Helvetica/Times/Courier)
```
Comment (kortschak) [inline vg/vgsvg/vgsvg.go:412]: Initial capital and final period.

File: vg/vgsvg/font_example_test.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	"gonum.org/v1/plot"
+	"gonum.org/v1/plot/font"
+	"gonum.org/v1/plot/plotter"
+	"gonum.org/v1/plot/vg"
+	"gonum.org/v1/plot/vg/draw"
+	"gonum.org/v1/plot/vg/vgsvg"
+)
+
+func Example_embedFonts() {
+	// use Latin-Modern fonts.
```
Comment (kortschak) [inline vg/vgsvg/font_example_test.go:25]: Initial capital.

**Review by kortschak — COMMENTED** [review]
Sorry, I missed these, but the golden files need regeneration anyway.

**Comment by sbinet** [comment]
PTAL.

I've added a simple mechanism that should preserve the old behaviour when relying on the `Liberation <==> Times/Helvetica/Courier` association map.
could you confirm?

**Comment by kortschak** [comment]
No, that doesn't fix the issue.

**Comment by sbinet** [comment]
(sorry, comments/PRs crossed each others... PS4 should address your latest comments)

**Comment by sbinet** [comment]
ok. I'll remove the mechanism and add a comment linking to the example, then.

where should I hang it?
on `vgsvg.Canvas`, `vgsvg.New` or at the package level?

**Comment by kortschak** [comment]
I would suggest that there also be documentation for installation of the liberation fonts, both in the package-level docs and in a README for the package.

**Comment by sbinet** [comment]
installing fonts is widely OS (flavour) dependent... (and I actually don't know how to install fonts on darwin nor Windows nor *BSDs...)
I've just redirected to the wikipedia page for now.

(alternatively, we could switch to "embed-fonts=true", like for PDFs: this would make `vgsvg` do the "right thing" by default.)

**Comment by kortschak** [comment]
Yeah, I think just embedding it probably the right thing to do. I just checked my install and I do have them installed, so even then we can't guaranteed the correct behaviour.

**Comment by sbinet** [comment]
PTAL

**Review by kortschak — APPROVED** [review]
The fonts in the originally incorrectly rendering fonts are still incorrectly rendering, but this has gone on long enough, so I'll approve.

**Comment by sbinet** [comment]
ok. thanks.

out of curiosity, what's your setup? (browser, OS)

**Comment by sbinet** [comment]
(the force-push is just to lump the `PSxxx` commits together)

**Comment by kortschak** [comment]
ff89.0.1 on ubuntu 18.04.4.

**Review by kortschak — APPROVED** [review]

**Comment by sbinet** [comment]
ah!
I can reproduce with ff89 on my archlinux system.

it seems that the culprit is specifying `font-family: Liberation, serif;` or `font-family: Liberation, serif, Times;` instead of, say: `font-family: Liberation, Times;`

let me see if there isn't some deeper flaw at work...

**Comment by sbinet** [comment]
aha!
for firefox (but it also works on my Chromium browser), it should read:
`font-family: Liberation Serif;`
(or all lower case: `font-family: liberation serif;`)
it would otherwise switch to its default font "DejaVu" which obviously doesn't metrically match with Liberation/Times.

I'll fix that.
but perhaps we could then switch back the default to "embed=false" ?

**Comment by kortschak** [comment]
> but perhaps we could then switch back the default to "embed=false" ?

If it work with that, then SGTM

**Comment by sbinet** [comment]
PTAL.

now `svgFontDescr` will generate the correct `font-family` name directly from the underlying raw SFNT data (using `sfnt.NameIDFamily`) and handle the mismatch of meaning for `font-variant` between SVG and gonum/plot/font.

**Comment by sbinet** [comment]
PTAL (just rebased off the new `master`)

**Review by kortschak — APPROVED** [review]
:tada:

**Pull request merged**